README.md
<h1 align="center" style="">SudoEvade</h1>

<p align="center">
   Execute Bash commands with root privileges, without entering a password or editing your Sudoers file.

</p>

<p align="center">
    <a href="">
       <img alt="MacOS" src="https://img.shields.io/badge/MacOS-x86/arm64-red.svg"/>
    </a>
    <a href="">
       <img alt="Linux" src="https://img.shields.io/badge/Linux-Ubuntu/Debian+-violet.svg"/>
    </a>
    <a href="https://github.com/BitesPotatoBacks/SudoEvade/releases">
        <img alt="Releases" src="https://img.shields.io/github/release/BitesPotatoBacks/SudoEvade.svg"/>
    </a>
    <a href="https://github.com/BitesPotatoBacks/SudoEvade/blob/main/LICENSE">
        <img alt="License" src="https://img.shields.io/github/license/BitesPotatoBacks/SudoEvade.svg"/>
    </a>
<!--     <a href="https://cash.app/$bitespotatobacks">
        <img alt="License" src="https://img.shields.io/badge/donate-Cash_App-default.svg"/>
    </a> -->
    <br>
</p>

## Project Deetds
### How the Command Works
In order to run commands as root without the usual requirements of `bash`, SudoEvade leverages the power of a launch Daemon.

1. The client binary captures your inputted command string and saves it to a file.
2. When that file changes, the Daemon Helper may read it and find the location of the command's binary.
3. Once found, the Daemon Helper clones the binary to a hidden directory and modifies the clone to elevate it's privileges.
4. Once the cloned binary is finished, the client can then execute your inputted command using the cloned and modified binary.

There are easier ways SudoEvade could be implemented, but I specifically went this route because I thought it would be more interesting to deal with.

### Usage Warning
If you mess something up in your system running a command as root under SudoEvade when you **should not** have, don't blame me for your lost files, broken OS install, or whatever...

## Installation and Usage
1. Download the .zip file from the [latest release](https://github.com/BitesPotatoBacks/SudoEvade/releases).
2. Unzip the .zip file and run the `install.sh` script in your terminal, like so: `sudo bash PATH/TO/SCRIPT/install.sh -i`. To see all installer options, use arg `-h`.
3. Once the installation is complete, you may execute a command with root priveleges using `sudoev`. To see all runtime options, use arg `-h`.

To check that SudoEvade is working properly, run `sudoev id -u`. If all is well, it should return a value of `0`.
___

If the install script fails and reports `Daemon did not start`, run `sudo bash PATH/TO/SCRIPT/install.sh -u` and then reinstall. If this fails, you may need to start the Daemon manually using one of the following command sets.
<details>
   
<summary>MacOS Manual Daemon Starting</summary>
   
```
sudo chmod 600 /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist
sudo launchctl load -w /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist
sudo launchctl start -w /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist
```
   
   Make sure the Daemon is running by checking `sudo launchctl list | grep "com.bitespotatobacks.SudoEvade"`.. 

</details>

<details>

<summary>Linux Manual Daemon Starting</summary>
   
```
sudo chmod 664 /etc/systemd/system/com.bitespotatobacks.SudoEvade.service
sudo systemctl daemon-reload
sudo systemctl start com.bitespotatobacks.SudoEvade
sudo systemctl enable com.bitespotatobacks.SudoEvade
 ```
   
   Make sure the Daemon is running by checking `systemctl | grep "com.bitespotatobacks.SudoEvade"`. 
   
</details>
  
## Issues
SudoEvade may suffer from issues when using certain shell builtins (like `kill`) or commands that require access to system groups (like `chown root:wheel` in macOS). Some of these problems can be worked around by using the tty piping arg `-t`.

If any other bugs or issues are identified or you want your system supported, please let me know in the [issues](https://github.com/BitesPotatoBacks/SudoEvade/issues) section.

## Support
If you would like to support this project, a small donation to my [Cash App](https://cash.app/$bitespotatobacks) would be much appreciated!



CHANGELOG.md
## v0.3.1 (July 31, 2022)
### Release Notes
This patch contains a crucial bug fixe for the installer and helper error recognition improvements. 

### Changelog
**Bug Fixes**
- Fixed install script not actually adding `/usr/local/bin` to PATH when missing (which could prevent the user from using the tool)

**Improvements**
- Code density improvements
- Improved recognition of helper failure by adding a unresponsive helper timeout (modify with `-T`)

___

## v0.3.0 (July 17, 2022)
### Release Notes
This minor contains crucial fixes and cleanups from the last version, as well as a couple of new features.

### Changelog
**Features**
- Added client command line argument to use tty redirection over binary cloning (experimental)
- Added installer command line argument to reinstall  

**Bug Fixes**
- Fixed running command from helper not accepting full command string args
- Fixed client command line args breaking command string input when separated
- (macOS) Fixed `chown` possibly working incorrectly by making use  of options `-t` or `-b`

**Improvements**
- Hide ugly install script errors and replaced with meaningful ones

___

## v0.2.0 (July 15, 2022)
### Release Notes
This minor contains brand new command line arguments to the client, as well as general improvements and minor fixes. More fixes and improvements to come in a future patch.

### Changelog
**Features**
- Added command line argument to run command in the background via the helper
- Added command line argument to attempt to force a command to run, ignoring all error and safety checks
- Added command line argument to run command rootless on helper failure

**Bug Fixes**
- (MacOS) Hopeful fix for install script skipping `chmod` for daemon plist randomly

**Improvements**
- Made scripts universal for all OS' rather than divided
- Improved on installer and helper outputs
- Hiding unneeded error from `rm` after command failure on exit

___

## v0.1.3 (June 26, 2022)
### Release Notes
This patch contains a crucial bug fix for macOS and general improvements for the tool.
### Changelog
**Bug Fixes**
- (MacOS) Fix for client failing to detect helper errors after system restart due to stderr log files being stored in `/tmp`

**Improvements**
- Helper may now report it's version with opt `-v`
- Install Script and Helper stdout have been further improved
- Generic dead code cleaning

___

## v0.1.2 (June 15, 2022)
### Release Notes
This patch contains crucial bug fixes and improvements as well as **Linux support** (official support for Ubuntu and Debian based distros).
### Changelog
**Bug Fixes**
- Fix for shell builtin commands failing to execute
- Fix for shell scripts not working properly when using shell builtin commands
- Fix for issue cause by previous patch that prevented root privileges on certain commands

**Improvements**
- Added version reporting to installer
- Improvements to error catching in stdout for helper diagnostic logs
- Generic efficiency and speed improvements

___

## v0.1.1 (June 12, 2022)
### Release Notes

This patch contains crucial bug fixes and improvements. More bug fixes and Linux support coming soon in upcoming releases. Make sure to uninstall your previous release before installing this patch.

### Changelog

**Bug Fixes**

- Fixed binaries located at paths containing spaces unable to execute
- Fixed binaries located at paths using trailing characters (./) unable to execute
- Fixed binaries reporting their names to contain a sudoev_ prefix

**Improvements**

- Helper tool sampling interval increased
- Helper tool diagnostic stdout simplified and stored in .log file rather than plain text

___

## v0.1.0 (June 5, 2022)
Initial release. Bug fixes and Linux support coming soon in upcoming releases.


install.sh
#!/bin/bash

#  install.sh
#  SudoEvade (Install Script)
#
#  Created by BitesPotatoBacks on 5/31/22.
#  Copyright (c) 2022 BitesPotatoBacks. All rights reserved.
#

OS=$(uname)
VER="v0.3.1"

RED=$(tput setaf 1)
GREEN=$(tput setaf 2)
PURPLE=$(tput setaf 5)

BOLD=$(tput bold)
NF=$(tput sgr0)

DIRSTUB="/Library/Caches/com.bitespotatobacks.SudoEvade"
[[ $OS == "Linux" ]] && DIRSTUB="/lib/com.bitespotatobacks.SudoEvade"
    
CMDFILE="$DIRSTUB/com.bitespotatobacks.SudoEvade.cmd.txt"
CMDARGFILE="$DIRSTUB/com.bitespotatobacks.SudoEvade.cmdarg.txt"
TTYFILE="$DIRSTUB/com.bitespotatobacks.SudoEvade.tty.txt"
HELPERFILE="$DIRSTUB/com.bitespotatobacks.SudoEvade.helper.txt"

DIR="$(dirname "${BASH_SOURCE[0]}")/src"

ROOT_GRP="root:wheel"

[[ "$OS" == "Linux" ]] && ROOT_GRP="root"

check_sudo() {
    if [ $UID -ne 0 ]; then
        echo "This script requires root privileges to install."
        echo "${BOLD}Re-run using Sudo to Install/Uninstall SudoEvade.${NF}"
        exit 1
    fi
}

install() {
    check_sudo

    echo " [1/6] Generating Directories"

    mkdir $DIRSTUB
    sudo chown $ROOT_GRP $DIRSTUB
    sudo chmod 4777 $DIRSTUB
    
    echo "" > $CMDFILE
    sudo chown $ROOT_GRP $CMDFILE
    sudo chmod 4777 $CMDFILE
    
    echo "" > $CMDARGFILE
    sudo chown $ROOT_GRP $CMDARGFILE
    sudo chmod 4777 $CMDARGFILE
    
    echo "" > $TTYFILE
    sudo chown $ROOT_GRP $TTYFILE
    sudo chmod 4777 $TTYFILE
    
    echo "" > $HELPERFILE
    sudo chown $ROOT_GRP $HELPERFILE
    sudo chmod 4777 $HELPERFILE

    echo " [2/6] Fixing Permissions"
    if [[ "$OS" == "Darwin" ]]; then
        if echo $PATH | grep -q "usr/local/bin"; then :; else
            printf "\n\nexport PATH=\"/usr/local/bin:\$PATH\"" >> ~/.zshrc
        fi
    fi
    
    chmod 755 $DIR/com.bitespotatobacks.SudoEvade.sh
    chmod 755 $DIR/sudoev.sh

    echo " [3/6] Relocating Client"
    cp $DIR/sudoev.sh $DIR/sudoev
    
    if [[ "$OS" == "Linux" ]]; then
        sudo mv $DIR/sudoev /usr/bin/sudoev
    else
        sudo mv $DIR/sudoev /usr/local/bin/sudoev
    fi

    echo " [4/6] Creating Daemon"
    if [[ "$OS" == "Linux" ]]; then
        sudo touch /etc/systemd/system/com.bitespotatobacks.SudoEvade.service &
        sudo chmod 664 /etc/systemd/system/com.bitespotatobacks.SudoEvade.service &
        sudo cat $DIR/com.bitespotatobacks.SudoEvade.service > /etc/systemd/system/com.bitespotatobacks.SudoEvade.service &
    else
        sudo touch /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist &
        sudo chmod 4777 /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist &
        sudo cat $DIR/com.bitespotatobacks.SudoEvade.plist > /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist &
    fi

    echo " [5/6] Creating Helper Tool"
    if [[ "$OS" == "Linux" ]]; then
        sudo touch /usr/bin/com.bitespotatobacks.SudoEvade.sh &
        sudo chmod 4777 /usr/bin/com.bitespotatobacks.SudoEvade.sh &
        sudo cat $DIR/com.bitespotatobacks.SudoEvade.sh > /usr/bin/com.bitespotatobacks.SudoEvade.sh &
        sudo chmod 4755 /usr/bin/com.bitespotatobacks.SudoEvade.sh &
    else
        sudo touch /Library/PrivilegedHelperTools/com.bitespotatobacks.SudoEvade.sh &
        sudo chmod 4777 /Library/PrivilegedHelperTools/com.bitespotatobacks.SudoEvade.sh &
        sudo cat $DIR/com.bitespotatobacks.SudoEvade.sh > /Library/PrivilegedHelperTools/com.bitespotatobacks.SudoEvade.sh &
        sudo chmod 4755 /Library/PrivilegedHelperTools/com.bitespotatobacks.SudoEvade.sh &
    fi

    echo " [6/6] Loading Daemon"
    if [[ "$OS" == "Linux" ]]; then
        sudo systemctl daemon-reload &
        sudo systemctl start com.bitespotatobacks.SudoEvade
        sudo systemctl enable com.bitespotatobacks.SudoEvade
        
        if systemctl | grep -q "com.bitespotatobacks.SudoEvade"; then
            echo "${GREEN}Install Complete!${NF}"
        else
            echo "${RED}Install Failed:${NF} Daemon did not start, please attempt reinstall"
        fi
    else
        while [[ $(ls -l /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist | awk '{print $1}') != "-rw-------" ]]; do
            sudo chmod 600 /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist
        done
        
        sudo launchctl load -w /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist &
        sudo launchctl start -w /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist
        
        if sudo launchctl list | grep -q "com.bitespotatobacks.SudoEvade"; then
            echo "${GREEN}Install Complete!${NF}"
        else
            echo "${RED}Install Failed:${NF} Daemon did not start, please attempt reinstall"
        fi
    fi
}

uninstall() {
    if [ ! -d $DIRSTUB ]; then
        echo "${RED}Uninstall Failed:${NF} Not previously installed ";
        exit
    fi
    
    check_sudo

    echo " [1/4] Removing Directories"
    sudo rm -r $DIRSTUB

    echo " [2/4] Removing Client"
    if [[ "$OS" == "Linux" ]]; then
        sudo rm /usr/bin/sudoev
    else
        sudo rm /usr/local/bin/sudoev
    fi

    echo " [3/4] Removing Daemon"
    if [[ "$OS" == "Linux" ]]; then
        sudo systemctl stop com.bitespotatobacks.SudoEvade &
        sudo rm /etc/systemd/system/com.bitespotatobacks.SudoEvade.service &
        sudo systemctl daemon-reload
    else
        sudo launchctl stop -w /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist &
        sudo launchctl unload -w /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist
        sudo rm /Library/LaunchDaemons/com.bitespotatobacks.SudoEvade.plist
    fi

    echo " [4/4] Removing Helper Tool"
    if [[ "$OS" == "Linux" ]]; then
        sudo rm /usr/bin/com.bitespotatobacks.SudoEvade.sh
        
        if systemctl | grep -q "com.bitespotatobacks.SudoEvade"; then
            echo "${RED}Uninstall Failed:${NF} Could not remove daemon"
        else
            echo "${PURPLE}Uninstall Complete!${NF}"
        fi
    else
        sudo rm /Library/PrivilegedHelperTools/com.bitespotatobacks.SudoEvade.sh
        
        if sudo launchctl list | grep -q "com.bitespotatobacks.SudoEvade"; then
            echo "${RED}Uninstall Failed:${NF} Could not remove daemon"
        else
            echo "${PURPLE}Uninstall Complete!${NF}"
        fi
    fi
}

usage() {
    echo "usage: install.sh -h | -V"
    echo "usage: install.sh [-i|-u|-r]"
    echo "Options: "
    echo "   -h   print help and exit"
    echo "   -V   print install version and exit"
    echo "   -i   install SudoEvade"
    echo "   -u   uninstall SudoEvade"
    echo "   -r   reinstall SudoEvade"
    exit
}

while getopts "uirhvV" OPT; do
    case $OPT in
        u) echo "${BOLD}Uninstalling SudoEvade ($OS $VER)${NF}";
           uninstall 2> /dev/null;
           exit ;;
        i) echo "${BOLD}Installing SudoEvade ($OS $VER) ${NF}";
           install 2> /dev/null;
           exit ;;
        r) echo "${BOLD}Reinstalling SudoEvade ($OS $VER)${NF}";
           uninstall 2> /dev/null; install 2> /dev/null;
           exit ;;
        v|V) echo "${BOLD}Installing:${NF} SudoEvade ($OS $VER)" ;;
        h|*) usage;;
    esac
done

[[ $# -eq 0 ]] && usage
